noir_stdlib/src/hash/sha256.nr

```diff
@@ -169,99 +169,352 @@ unconstrained fn build_msg_block<let N: u32>(
 fn verify_msg_block<let N: u32>(
     msg: [u8; N],
     message_size: u32,
     msg_block: MSG_BLOCK,
     msg_start: u32,
 ) -> BLOCK_BYTE_PTR {
-    let mut msg_byte_ptr: u32 = 0; // Message byte pointer
+    let mut msg_byte_ptr = 0;
     let mut msg_end = msg_start + BLOCK_SIZE;
     if msg_end > N {
         msg_end = N;
     }
+    // We might have to go beyond the input to pad the fields.
+    if msg_end % INT_SIZE != 0 {
+        msg_end = msg_end + INT_SIZE - msg_end % INT_SIZE;
+    }
 
-    for k in msg_start..msg_end {
-        if k < message_size {
-            assert_eq(msg_block[msg_byte_ptr], msg[k]);
+    // Reconstructed packed item.
+    let mut msg_item: u32 = 0;
+
+    // Inclusive at the end so that we can compare the last item.
+    let mut i: u32 = 0;
+    for k in msg_start..=msg_end {
+        if k % INT_SIZE == 0 {
+            // If we consumed some input we can compare against the block.
+            if msg_start < message_size & k > msg_start {
+                assert_eq(msg_block[i], msg_item as u32);
+                i = i + 1;
+                msg_item = 0;
+            }
+        }
+        // Shift the accumulator
+        msg_item = lshift8(msg_item, 1);
+        // If we have input to consume, add it at the rightmost position.
+        if k < message_size & k < msg_end {
+            msg_item = msg_item + msg[k] as u32;
             msg_byte_ptr = msg_byte_ptr + 1;
         }
     }
 
     msg_byte_ptr
 }
 
-// Verify the block we are compressing was appropriately padded with zeroes by `build_msg_block`.
+// Verify the block we are compressing was appropriately padded with zeros by `build_msg_block`.
 // This is only relevant for the last, potentially partially filled block.
 fn verify_msg_block_padding(msg_block: MSG_BLOCK, msg_byte_ptr: BLOCK_BYTE_PTR) {
+    // Check all the way to the end of the block.
+    verify_msg_block_zeros(msg_block, msg_byte_ptr, INT_BLOCK_SIZE);
+}
+
+// Verify that a region of ints in the message block are (partially) zeroed,
+// up to an (exclusive) maximum which can either be the end of the block
+// or just where the size is to be written.
+fn verify_msg_block_zeros(
+    msg_block: MSG_BLOCK,
+    mut msg_byte_ptr: BLOCK_BYTE_PTR,
+    max_int_byte_ptr: u32,
+) {
     // This variable is used to get around the compiler under-constrained check giving a warning.
     // We want to check against a constant zero, but if it does not come from the circuit inputs
     // or return values the compiler check will issue a warning.
     let zero = msg_block[0] - msg_block[0];
 
-    for i in 0..BLOCK_SIZE {
-        if i >= msg_byte_ptr {
+    // First integer which is supposed to be (partially) zero.
+    let mut int_byte_ptr = msg_byte_ptr / INT_SIZE;
+
+    // Check partial zeros.
+    let modulo = msg_byte_ptr % INT_SIZE;
+    if modulo != 0 {
+        let num_zeros = INT_SIZE - modulo;
+        let mask = if num_zeros == 3 {
+            256 * 256 * 256
```

Comment:
Could make these globals to avoid the magic numbers

Reply:
Added TWO_POW... globals.